Crash in `SavedMatrix#finished`: IllegalStateException: step.testExecutionStep.testTiming must not be null

We encountered this crash using Flank version v4.2.0, while running 75 shards.

It looks like there's a null check for `step.testExecutionStep`, but in this case `step.testExecutionStep.testTiming` was `null`.

```
10:42:50   4m 48s matrix-285i5yv89ulkl RUNNING
10:43:47   5m 43s matrix-285i5yv89ulkl Non-retryable error occurred during results processing.
10:52:53 Request failed, retrying 1x java.net.SocketTimeoutException: Read timed out
11:02:45   24m 38s matrix-285i5yv89ulkl FINISHED
11:02:45 Exception in thread "main" picocli.CommandLine$ExecutionException: Error while running command (ftl.cli.firebase.test.android.AndroidRunCommand@6d367020): java.lang.IllegalStateException: step.testExecutionStep.testTiming must not be null
11:02:45     at picocli.CommandLine.execute(CommandLine.java:1168)
11:02:45     at picocli.CommandLine.access$800(CommandLine.java:141)
11:02:45     at picocli.CommandLine$RunLast.handle(CommandLine.java:1367)
11:02:45     at picocli.CommandLine$RunLast.handle(CommandLine.java:1335)
11:02:45     at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1243)
11:02:45     at picocli.CommandLine.parseWithHandlers(CommandLine.java:1526)
11:02:45     at picocli.CommandLine.run(CommandLine.java:1974)
11:02:45     at picocli.CommandLine.run(CommandLine.java:1920)
11:02:45     at ftl.Main$Companion.main(Main.kt:48)
11:02:45     at ftl.Main.main(Main.kt)
11:02:45 Caused by: java.lang.IllegalStateException: step.testExecutionStep.testTiming must not be null
11:02:45     at ftl.json.SavedMatrix.finished(SavedMatrix.kt:81)
11:02:45     at ftl.json.SavedMatrix.update(SavedMatrix.kt:57)
11:02:45     at ftl.run.TestRunner.pollMatrices(TestRunner.kt:249)
11:02:45     at ftl.run.TestRunner.newRun(TestRunner.kt:353)
11:02:45     at ftl.run.TestRunner$newRun$1.invokeSuspend(TestRunner.kt)
```

--- a/release_notes.md
+++ b/release_notes.md
@@ -1,5 +1,6 @@
 ## v4.4.0 (unreleased)
 
+- [#501](https://github.com/TestArmada/flank/pull/501) Fix nullability check in SavedMatrix. ([bootstraponline](https://github.com/bootstraponline))
 - [#493](https://github.com/TestArmada/flank/pull/493) Fix getDefaultBucket timeout. ([bootstraponline](https://github.com/bootstraponline))
 
 ## v4.3.1

--- a/test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt
+++ b/test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt
@@ -77,8 +77,11 @@ class SavedMatrix(matrix: TestMatrix) {
             val step = GcToolResults.getResults(it.toolResultsStep)
             updateOutcome(step.outcome)
 
-            if (step.testExecutionStep == null) return
-            val billableMinutes = Billing.billableMinutes(step.testExecutionStep.testTiming.testProcessDuration.seconds)
+            // testExecutionStep, testTiming, etc. can all be null.
+            // sometimes testExecutionStep is present and testTiming is null
+            val testTimeSeconds = step.testExecutionStep?.testTiming?.testProcessDuration?.seconds ?: return
+
+            val billableMinutes = Billing.billableMinutes(testTimeSeconds)
 
             if (AndroidCatalog.isVirtualDevice(it.environment?.androidDevice)) {
                 billableVirtualMinutes += billableMinutes

--- a/test_runner/src/test/kotlin/ftl/test/util/TestArtifact.kt
+++ b/test_runner/src/test/kotlin/ftl/test/util/TestArtifact.kt
@@ -74,9 +74,11 @@ object TestArtifact {
 
     private fun remoteAssetLink(): Element {
         val doc = getHtml()
-        val downloadLinks = doc.select("li.d-block > a")
+        val downloadLinks = doc.select("a")
+        // /Flank/test_artifacts/releases/download/latest/2552E672A1366D1B3A06452AA7256ABC.zip
+        val zipRegex = Regex(".*releases/download/latest/\\H{32}\\.zip$")
         return downloadLinks.find {
-            it.attr("href").endsWith(".zip") && it.attr("href").contains("releases/download")
+            it.attr("href").matches(zipRegex)
         } ?: throw RuntimeException("Download link not found in html!")
     }
 

